plutarch-test/src/Plutarch/Test/Golden.hs

```diff
@@ -0,0 +1,184 @@
+import GHC.Stack (HasCallStack)
+import System.FilePath ((</>))
+import Test.Syd (
+  Expectation,
+  Spec,
+  TestDefM,
+  describe,
+  getTestDescriptionPath,
+  it,
+  pureGoldenTextFile,
+ )
+
+import Plutarch
+import Plutarch.Benchmark (benchmarkScript')
+import Plutarch.Evaluate (evaluateScript)
+import Plutarch.Internal (Term (Term, asRawTerm))
+import Plutarch.Test.ListSyntax (ListSyntax, listSyntaxAdd, listSyntaxAddSubList, runListSyntax)
+import qualified Plutus.V1.Ledger.Scripts as Scripts
+
+data GoldenValue = GoldenValue
+  { goldenValueUplcPreEval :: Text
+  , goldenValueUplcPostEval :: Text
+  , goldenValueBench :: Text
+  , goldenValueExpectation :: Maybe Expectation
+  }
+
+{- | Class of types that represent `GoldenValue`
+
+  This class exists for syntatic sugar provided by (@->) (via `TermExpectation`).
+-}
+class HasGoldenValue a where
+  mkGoldenValue :: a -> GoldenValue
+
+mkGoldenValue' :: ClosedTerm a -> Maybe Expectation -> GoldenValue
+mkGoldenValue' p mexp =
+  GoldenValue
+    (T.pack $ printScript $ compileD p)
+    (T.pack $ printScript $ evaluateScriptAlways $ compileD p)
+    (TL.toStrict $ Aeson.encodeToLazyText $ benchmarkScript' $ compileD p)
+    mexp
+
+-- We derive for `Term s a` only because GHC prevents us from deriving for
+-- `ClosedTerm a`. In practice, this instance should be used only for closed
+-- terms.
+instance HasGoldenValue (Term s a) where
+  mkGoldenValue p = mkGoldenValue' (unsafeClosedTerm p) Nothing
+    where
+      -- Because, we need a function with this signature.
+      unsafeClosedTerm :: Term s a -> ClosedTerm a
+      unsafeClosedTerm t = Term $ asRawTerm t
+
+{- | A `Term` paired with its evaluation expectation
+
+  Example:
+  >>> TermExpectation (pcon PTrue) $ \p -> pshouldBe (pcon PTrue)
+-}
+data TermExpectation a = TermExpectation (ClosedTerm a) (ClosedTerm a -> Expectation)
+
+-- | Test an expectation on a golden Plutarch program
+(@->) :: ClosedTerm a -> (ClosedTerm a -> Expectation) -> TermExpectation a
+(@->) p f = TermExpectation p (\p' -> f p')
+
+infixr 1 @->
+
+instance HasGoldenValue (TermExpectation a) where
+  mkGoldenValue (TermExpectation p f) = mkGoldenValue' p (Just $ f p)
+
+-- | The key used in the .golden files containing multiple golden values
+newtype GoldenKey = GoldenKey Text
+  deriving newtype (Eq, Show, Ord, IsString)
+
+goldenKeyString :: GoldenKey -> String
+goldenKeyString (GoldenKey s) = T.unpack s
+
+instance Semigroup GoldenKey where
+  GoldenKey s1 <> GoldenKey s2 = GoldenKey $ s1 <> "." <> s2
+
+goldenPath :: FilePath -> GoldenKey -> FilePath
+goldenPath baseDir (GoldenKey k) =
+  baseDir </> T.unpack k <> ".golden"
+
+-- | Group multiple goldens values in the same file
+combineGoldens :: [(GoldenKey, Text)] -> Text
+combineGoldens xs =
+  T.intercalate "\n" $
+    (\(GoldenKey k, v) -> k <> " " <> v) <$> xs
+
+-- | Specify goldens for the given Plutarch program
+(@|) :: HasGoldenValue v => GoldenKey -> v -> ListSyntax (GoldenKey, GoldenValue)
```

Comment:
So this is still unsafe unfortunately. `v` could be a `Term s a` that references free variables.
You could split this up into two functions, one for `ClosedTerm a` and one for `TermExpectation a`.
Another possible solution is to add the `s` to `HasGoldenValue`, such that it is `(forall s. HasGoldenValue s v) =>`. The instance for `TermExpectation` would not make use of the `s`.

Reply:
```haskell
bad' :: forall s a. GoldenKey -> Term s a -> ListSyntax (GoldenKey, GoldenValue)
bad' k v = listSyntaxAdd (k, mkGoldenValue @s v)

xc :: GoldenKey -> (ListSyntax (GoldenKey, GoldenValue) -> Term s a) -> Term s (POpaque :--> a)
xc key f = plam $ \(x :: Term _ POpaque) -> f (bad' key x)
```

Reply:
This is quite obviously nonsense.

Reply:
I fixed this myself.